src/hotspot/share/utilities/resourceHash.hpp

```diff
@@ -193,17 +188,20 @@ class ResourceHashtableBase : public STORAGE {
     return false;
   }
 
-  // ITER contains bool do_entry(K const&, V const&), which will be
+  // bool ITER::do_entry(K const& key, V& value) will be
   // called for each entry in the table.  If do_entry() returns false,
   // the iteration is cancelled.
+  //
+  // do_entry() may modify the value. However, it should not modify the key,
+  // or else the table may no longer be properly hashed.
   template<class ITER>
-  void iterate(ITER* iter) const {
-    Node* const* bucket = table();
+  void iterate(ITER* iter) {
+    Node** bucket = table();
     const unsigned sz = table_size();
     while (bucket < bucket_at(sz)) {
       Node* node = *bucket;
       while (node != NULL) {
-        bool cont = iter->do_entry(node->_key, node->_value);
+        bool cont = iter->do_entry(const_cast<K const&>(node->_key), node->_value);
```

Comment:
It seems to me that iterate should be const because it doesn't change the table.  Whether or not it changes the pointers in the Nodes is not part of the contract, since they could have been declared non-const.